cleanlab/datalab/internal/issue_manager/underperf_group.py

```diff
@@ -0,0 +1,389 @@
+            # with the provided features
+            knn_graph = None
+
+        return knn_graph
+
+    def collect_info(
+        self,
+        knn_graph: csr_matrix,
+        n_clusters: int,
+        cluster_ids: npt.NDArray[np.int_],
+        performed_clustering: bool,
+        worst_cluster_id: int,
+    ) -> Dict[str, Any]:
+        params_dict = {
+            "k": self.k,
+            "metric": self.metric,
+            "threshold": self.threshold,
+        }
+
+        knn_info_dict = {}
+        if knn_graph is not None:
+            N = knn_graph.shape[0]
+            dists = knn_graph.data.reshape(N, -1)[:, 0]
+            nn_ids = knn_graph.indices.reshape(N, -1)[:, 0]
+
+            knn_info_dict = {
+                "nearest_neighbor": nn_ids.tolist(),
+                "distance_to_nearest_neighbor": dists.tolist(),
+            }
+        statistics_dict = self._build_statistics_dictionary(knn_graph=knn_graph)
+
+        cluster_stat_dict = self._get_cluster_statistics(
+            n_clusters=n_clusters,
+            cluster_ids=cluster_ids,
+            performed_clustering=performed_clustering,
+            worst_cluster_id=worst_cluster_id,
+        )
+        info_dict = {
+            **params_dict,
+            **knn_info_dict,
+            **statistics_dict,
+            **cluster_stat_dict,
+        }
+
+        return info_dict
+
+    def _build_statistics_dictionary(self, knn_graph: csr_matrix) -> Dict[str, Dict[str, Any]]:
+        statistics_dict: Dict[str, Dict[str, Any]] = {"statistics": {}}
+
+        # Add the knn graph as a statistic if necessary
+        graph_key = "weighted_knn_graph"
+        old_knn_graph = self.datalab.get_info("statistics").get(graph_key, None)
+        old_graph_exists = old_knn_graph is not None
+        prefer_new_graph = (
+            not old_graph_exists
+            or (isinstance(knn_graph, csr_matrix) and knn_graph.nnz > old_knn_graph.nnz)
+            or self.metric != self.datalab.get_info("statistics").get("knn_metric", None)
+        )
+        if prefer_new_graph:
+            if knn_graph is not None:
+                statistics_dict["statistics"][graph_key] = knn_graph
+                if self.metric is not None:
+                    statistics_dict["statistics"]["knn_metric"] = self.metric
+
+        return statistics_dict
+
+    def _get_cluster_statistics(
+        self,
+        n_clusters: int,
+        cluster_ids: npt.NDArray[np.int_],
+        performed_clustering: bool,
+        worst_cluster_id: int,
+    ) -> Dict[str, Dict[str, Any]]:
+        """Get relevant cluster statistics.
+
+        Args:
+            n_clusters (int): Number of clusters
+            cluster_ids (npt.NDArray[np.int_]): Cluster IDs for each datapoint.
+            performed_clustering (bool): If True, clustering was performed on
+            features passed to `find_issues`. If False, `cluster_ids` were explicitly
```

Comment:
Slight change of wording, make sure to adjust this to pass the formatting check.

```suggestion
            performed_clustering (bool): Set to True to indicate that clustering was performed on
            `features` passed to `find_issues`. Set to False to suggest that `cluster_ids` were explicitly
```